# Likninger og algoritmer

### Babylonsk rotalgoritme

Regner ut roten av et tall $ a $ ved å sette det inn i $ x = \frac{1}{2} * (x + \frac{a}{x}) $ flere ganger.

In [ ]:
x = 100
a = 12

for i in range(10):
    x = 0.5 * (x + a/x)

    print(x)

### Halveringsmetoden for å løse likninger

Finne nullpunktet i et intervall er akkurat det samme som å finne løsningen på en likning.

In [25]:
def f(x):
        return 3*x**5 - 2*x**2 - 8

def halveringsmetoden(a, b, f, tol = 10E-10):
    i = 0
    m = (a + b) / 2
    while abs(f(m)) > tol:
        if f(a) * f(m) < 0:
            b = m
        if f(b) * f(m) < 0:
            a = m
        m = (a + b) / 2
        i += 1
    return m, i

m, i = halveringsmetoden(-5, 10, f)
print("Nullpunktet er", m, "og løkka kjørte", i, "ganger.")

Nullpunktet er 1.3063395069002581 og løkka kjørte 37 ganger.


### Newtons metode

Vi har en kontinuerlig funksjon f. Vi tilnærmer nullpunktet til f med nullpunktet til tangentene til f(x).

$$ x_{n+1} = x_{n} - \frac{f(x_{n})}{f'(x_{n})} $$ 

Fordelen er at man kan ha mye større intervaller som blir utregnet mye fortere.
Ulempene er at algoritmen takler lokale topp- og bunnpunkter dårlig og den krever at man vet den deriverte.

In [35]:
def f(x):
    return x**2 - 4

def fder(x):
    return 2*x

def newtons(x, f, fder,tol = 10E-10, N=100):
    i = 0
    while abs(f(x)) > tol and i < N:
        x = x - f(x)/fder(x)
        i += 1
    return x, i

x, i = newtons(2500, f, fder)
print("Nullpunktet er", x, "og løkka kjørte", i, "ganger.")

Nullpunktet er 2.0000000000164926 og løkka kjørte 14 ganger.


### Oppgave "Skyte spurv med kanoner"
#### a)

Kulens høyde gitt en viss tid $y(t)$ er gitt ved ???

#### b)

Kulens strekning $x(t)$ er gitt ved $x(t)=v_{0} \cdot \cos{\theta} \cdot t$ fordi, hvis man ser for seg kulens start som en trekant, med en normal linje mot x-aksen, som viser strekningen, må man finne denne x-verdien, som er den nærmestliggende kateten til utskytningvinkelen.

#### c) d) e)

In [3]:
from pylab import math as m

g = 9.81
v0 = 18

def y(t, theta = m.pi/10):
    return v0 * m.sin(theta) * t - (1/2) * g * t**2

def yder(t, theta = m.pi/10):
    return v0 * m.sin(theta) - g * t

def x(t, theta):
    return v0 * m.cos(theta) * t

def newtons(f, fder, x, tol = 10E-10):
    i = 0
    while abs(f(x)) > tol and i < 100:
        x = x - f(x)/fder(x)
        i += 1
    return x, i

tid, i = newtons(y, yder, 20)
print("Tid: ", tid)

dist = x(tid, m.pi/10)
print("Strekning: ", dist)

# Oppgave e): Kjører gjennom en liste med forskjellige vinkler og finner distansen med x-funksjonen

thetalist = [m.pi/6, m.pi/5, m.pi/4, m.pi/3]

for theta_n in thetalist:
    print("For vinkel", theta_n, "radianer er avstanden", x(tid, theta_n), "meter.")

Tid:  1.1340073188072652
Strekning:  19.41309090140481
For vinkel 0.5235987755982988 radianer er avstanden 17.67742463096227
For vinkel 0.6283185307179586 radianer er avstanden 16.513761467891637
For vinkel 0.7853981633974483 radianer er avstanden 14.433556770788265
For vinkel 1.0471975511965976 radianer er avstanden 10.20606586926539


#### f)

Vi noterte på pcen til Nora og dessverre mistet vi alle underveisdataene, og beholdt kun de gjennomsnittlige verdiene. Ballen gikk i snitt 4.67 meter med en startfart på 5.83 m/s. Da skjøt vi med en vinkel (theta) på 30 deg eller 0.524 rad (tilsvarende $\frac{\pi}{6}$).

In [4]:
g = 9.81
v0 = 5.83

def y(t, theta = m.pi/6):
    return v0 * m.sin(theta) * t - (1/2) * g * t**2

def yder(t, theta = m.pi/6):
    return v0 * m.sin(theta) - g * t

def x(t, theta):
    return v0 * m.cos(theta) * t

def newtons(f, fder, x, tol = 10E-10):
    i = 0
    while abs(f(x)) > tol and i < 100:
        x = x - f(x)/fder(x)
        i += 1
    return x, i

tid, i = newtons(y, yder, 20)
print("Tid: ", tid)

dist = x(tid, m.pi/10)
print("Strekning: ", dist)

Tid:  0.5942915392456904
Strekning:  3.2951442229057677


Strekningen er feil, men det er å forvente, ettersom en slik teoretisk utregning ikke regner med variabler som luftmotstand. 